Collect pieces for SSP 2024/ScenarioMIP

khaeru opened the pull request:
This PR is to track the **working branch `ssp-dev`**, which exists to help with work on the SSP 2024/ScenarioMIP project by collecting or pointing to all the code that is actually used to execute the workflow.

The goal is to be complete and up-to-date as possible for that purpose; other aspects (code quality, cleanliness, or readiness for `main`) are less important.

## How to use

- Per #234, it is not necessary to use this branch. If you can make an atomic PR for individual, small changes, do that first. This branch will be regularly rebased on latest `main`, so will include those updates.
- Update this branch, e.g. by cherry-picking commits or copying code from various other working branches used for this project.
- Make a PR that targets into this branch, and then request that someone review it. Continue work on this branch, rather than the PR source branch.

## How to review

TBD: this PR may not eventually be merged.

## PR checklist

- ~Continuous integration checks all ✅~ TBD if this will be required.
- ~Add or expand tests; coverage checks both ✅~
- ~Add, expand, or update documentation.~
- ~Update doc/whatsnew.~

khaeru commented:
@GamzeUnlu95 @macflo8 —there now appears to be a conflict in a particular materials data file. I think this is the same one modified in #247 (FYI also @glatterf42 who reviewed there).

Can you please say whether these two changes to the file are perhaps the same, or different? Does one supersede the other, or are they parallel/non-overlapping?

This will allow to rebase and update the branch, which we need to do in order for it to be usable in the ScenarioMIP process.

khaeru commented:
> Can you please say whether these two changes to the file are perhaps the same, or different? Does one supersede the other, or are they parallel/non-overlapping?

The reply (in Slack) was that the changes in #247 supersede the ones that were on this branch.
Accordingly, I did:
- `git switch ssp-dev && git pull --rebase` update my local copy of the branch.
- `git rebase -i main` initiate interactive rebase.
- Chose `edit` for the commit with subject "Move co2_trans_disp parametrization to file" —the only commit on this branch that modified the file.
- Removed the changes to that file from the commit:
  ```
  git restore --staged message_ix_models/data/material/steel/Global_steel_MESSAGE.xlsx 
  git restore message_ix_models/data/material/steel/Global_steel_MESSAGE.xlsx 
  ```
- `git rebase --continue`
- `git push --force`

measrainsey commented on message_ix_models/data/costs/energy/cost_reduction.csv:
```diff
@@ -1,99 +1,102 @@
+wind_res2,Renewable,0.15,0.3,0.53,0.53,0.65,0.75
+wind_res3,Renewable,0.15,0.3,0.53,0.53,0.65,0.75
+wind_res4,Renewable,0.15,0.3,0.53,0.53,0.65,0.75
+dac_lt,CCS,0.1,0.3,0.5,0.7,0.9
```
gotcha! @ywpratama could you add `low_medium` values for the DAC technologies? since i wouldn't know the assumptions to make here

khaeru commented:
I see the branch is showing conflicts, so I will rebase manually and force-push.

khaeru commented:
> I see the branch is showing conflicts, so I will rebase manually and force-push.

Now done.

macflo8 commented:
I will drop my 3 most recent commits and force-push because they break the model build . The commits will be added later again when the industry revisions are fully functional.

khaeru commented:
Rebased after merge of #255.

khaeru commented:
- Rebased after merge of #225.
- Cherry-picked a single commit by @adrivinca from #261.

khaeru commented:
Rebased after the merge of #258.

In doing so, there was a conflict in message_ix_models/data/costs/materials/tech_map.csv. I resolved this in favour of the version on `main` per that other PR, which is more recent.

FYI @measrainsey if you are making a PR into `main` and happen to notice that it touches files also modified on this branch, please feel free to cherry-pick particular commits from this branch into a smaller PR branch. That way, they can be reviewed and merged, and then they will vanish from this branch when it is later rebased. In the long run this can shrink the work we must eventually do to incorporate the contents of this branch in `main`.

macflo8 commented:
After clarifying with @ywpratama, we figured out that the last two commits were pushed accidentally by him. I will drop them now and force-push.

khaeru commented:
I'm going to rebase the branch again on `main` to avoid the "Merge branch 'main' into ssp-dev" that we see at the end here. Should be done in about an hour; please no one touch the branch in the meantime.

khaeru commented on message_ix_models/model/material/data_util.py:
```diff
@@ -1012,55 +202,248 @@ def read_iea_tec_map(tec_map_fname: str) -> pd.DataFrame:
-def get_hist_act_data(
-    map_fname: str, years: list or None = None, iea_data_path: Optional[str] = None
+def read_sector_data(
+    scenario: message_ix.Scenario, sectname: str, ssp: str | None, filename: str
```
FYI @glatterf42 @macflo8 —for message-ix-models we provide support for all currently-maintained versions of Python. The earliest is currently Python 3.9. This syntax `str | None` is only supported from Python 3.10 onwards, so this line will cause the Python 3.9 CI jobs to fail. Until we drop support for Python 3.9, please use `Optional[str]`.

khaeru commented:
I will rebase this branch to incorporate #303 and resolve the conflicts. Please hold off on pushes in the meanwhile, @macflo8 and others.

khaeru commented:
> I will rebase this branch to incorporate #303 and resolve the conflicts. Please hold off on pushes in the meanwhile, @macflo8 and others.

Now done, please carry on!

khaeru commented:
After the merge of #300, GitHub claimed there were merge conflicts and did not allow an update-with-rebase through the web UI. I rebased the branch locally (simply `git rebase main`); this completed automatically without any conflicts to resolve. The rebased branch omits merge commits from #316 #317 #318, leaving only the commits from the respective branches

khaeru commented:
> After the merge of #300, GitHub claimed there were merge conflicts and did not allow an update-with-rebase through the web UI. I rebased the branch locally (simply `git rebase main`); this completed automatically without any conflicts to resolve.

Same happened after the merge of #325. It may be the branch is getting too long for GitHub to automatically rebase it through the web UI.

khaeru commented:
Rebased after merge of #395; the branch goes from 134 to 109 commits (-25).
